### ISSUE TITLE
Error build qz-tray 2.1 using JDK 1.8

### ISSUE THREAD

**Issue by Fermec28** [qzind/tray#329]
Hello everyone, i try to build qz tray version 2.1 on raspberry pi : 
java -version 
```bash
java version "1.8.0_172"
Java (TM) SE Runtime Environment (build 1.8.0_172-b11)
Java HotSpot(TM) Client VM (build 25.172-b11, mixed mode)
```
also i had downloaded and added  [JavaFX Embedded SDK ](http://gluonhq.com/products/mobile/javafxports/get/) following this [steps](http://docs.gluonhq.com/javafxports/#anchor-1) (2.1.4. Embedded)

But when i try to build with sudo  ant command, it show me this currently error:
```bash
compile-socket:
    [mkdir] Created dir: /home/pi/Desktop/tray/out/build/qz-tray
    [javac] Compiling 82 source files to /home/pi/Desktop/tray/out/build/qz-tray
    [javac] /home/pi/Desktop/tray/src/qz/printer/action/WebApp.java:8: error: package javafx.embed.swing does not exist
    [javac] import javafx.embed.swing.SwingFXUtils;
    [javac]                          ^
    [javac] /home/pi/Desktop/tray/src/qz/printer/action/WebApp.java:228: error: cannot find symbol
    [javac]                 capture.set(SwingFXUtils.fromFXImage(snapshot, null));
    [javac]                             ^
    [javac]   symbol:   variable SwingFXUtils
    [javac]   location: class WebApp
    [javac] 2 errors

BUILD FAILED
/home/pi/Desktop/tray/build.xml:39: Compile failed; see the compiler error output for details.

Total time: 9 seconds

```
so how i can resolve this error and build 2.1 version of qz ?

**Comment by tresf** [comment]
> so how i can resolve this error and build 2.1 version of qz ?

Install Java FX.

**Comment by tresf** [comment]
I believe JavaFX is getting harder to run on Pi, so we might need to add a compile-time flag to disable HTML support.

You could also compile on a regular desktop and copy the `.run`.  We don't do anything architecture-specific, so the `.run` for x86 should work just fine on any other architecture, such as ARM.

**Comment by blaggacao** [comment]
> You could also compile on a regular desktop and copy the .run. We don't do anything architecture-specific, so the .run for x86 should work just fine on any other architecture, such as ARM.

We just did that, and are now making tests to run it on a PI.
It would be quite interesting for our use case to have reliable PDF/HTML support on Raspberry PI with headless mode.

Am I right, that PDF/HTML goes together?

**Comment by tresf** [comment]
> Am I right assuming, that PDF/HTML goes together?

PDF should work.  It's 100% Apache PDFBOX under the covers; It should work headless as well.

HTML is the black sheep.  Not only does it use the -- somewhat buggy -- JavaFX framework but our attempts to run it headless have been unsuccessful.

As a workaround for HTML, there're some JavaScript libraries that will turn a DIV into a canvas object and export to PNG, but we've yet to find an alternative to JavaFX it doesn't purely Java-side.

**Comment by tresf** [comment]
> OK, it's not java, but works as a command line utility (and has a library in C): https://wkhtmltopdf.org/

That's an excellent library although we haven't made any attempt to shim in Pi support yet.  Ideally, the library that we use can communicate directly with the printer.  Any intermediary formats (PNG, PDF) would cause unnecessary overhead (as HTML support already does in the 2.0 branch).

You may be able to get your hands on an older JRE build before JavaFX was removed.  I'm not sure how reliable it will be.

**Comment by tresf** [comment]
### OpenJDK

For the Raspberry Pi, this will install OpenJDK (Java8u111) with JavaFX support directly on the Pi.
Note, Java also provides a JDK for the Pi, but it is missing JavaFX support.  We'll attempt that in a separate post. :)

**JavaFX**
```bash
# fixes "missing `server' JVM at `/usr/lib/jvm/java-8-openjdk-armhf/jre/lib/arm/server/libjvm.so'"
# fixes "cannot find symbol `javafx.embed.swing.SwingFXUtils;`"
sudo apt-get purge openjdk-8-jre-headless
sudo apt-get install openjdk-8-jre-headless
sudo apt-get install openjdk-8-jre
sudo apt-get install openjdk-8-jdk
sudo apt-get install openjfx
```

**Build Tools**
```bash
# javac is already installed as part of openjdk-8-jdk
sudo apt-get install ant makeself
```

**Compile, Install**
 ```bash
# git should come with the pi, if not, sudo apt-get install git
git clone https://github.com/qzind/tray
cd tray
# if ant complains about JAVA_HOME, make sure openjdk-8-jdk was installed :)
ant makeself
# note, certutil will be downloading at install if not installed
# certutil is used to get the pi's chromium browser to recognized the self-signed cert
sudo bash out/qz-tray-*.run -- -y
```

**Start QZ Tray**
```bash
# kill any running versions
pkill -9 java
# if using ssh/putty, export DISPLAY so Java and chromium can attach
# if not using ssh/putty, you can skip the export DISPLAY command
export DISPLAY=:0.0
# start chromium for first time to get certdb created
chromium-browser
# close chromium-browser manually
# start qz tray as bg process
nohup java -jar /opt/qz-tray/qz-tray.jar &
# monitor qz tray startup by reading terminal log
tail -f nohup.out
# should say "... started on ports 8181,8282"
```

**Test QZ Tray**
```bash
chromium-browser demo.qz.io
```

If you'd like to start using the Raspberry Pi as a print server, continue to https://qz.io/wiki/print-server.

**Test HTML Printing**
```bash
# optional, to see if the QZ Tray server can handle HTML
sudo apt-get install cups-pdf
# important: restart QZ Tray so that it can see the new printer
# assuming cups wasn't already installed, this will install cups and start it automatically
# a printer called "cups-pdf" will be available for test printing
# last, from sample.html, "Pixel Printing" tab:
#    search for "pdf" printer, click "Print HTML" button
```

<img width="635" alt="screen shot 2018-05-19 at 2 45 29 pm" src="https://user-images.githubusercontent.com/6345473/40271969-582caae0-5b73-11e8-9015-2d60befabdfe.png">

**Status**

Well, this fixes the compilation problems but unfortunately, [my JavaFX crashed](https://gist.github.com/tresf/59a6eef5ba2db429ec782b8a96c62a8a).  I believe there's a discrepancy between the `openjfx` runtime version and the base jre version as I've observed by the Java 8u111 versus the crash log stating Java 8u171.  Next, we'll try to get HTML support using Oracle JRE. :)

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0xa13326e0, pid=27480, tid=0xa806f470
#
# JRE version: OpenJDK Runtime Environment (8.0_171-b11) (build 1.8.0_171-8u171-b11-1~deb9u1-b11)
# Java VM: OpenJDK Client VM (25.171-b11 mixed mode linux-aarch32 )
# Problematic frame:
# C  [libjfxwebkit.so+0x14ca6e0]
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/pi/hs_err_pid27480.log
[TRACE] 2018-05-19 19:07:58,633 @ qz.printer.action.WebApp:?
	Waiting on capture..
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
```

**Comment by tresf** [comment]
Ok... so a few things I'm learning the hard way... 

1. Oracle Java 8 is already pre-installed on the PI but I've decided to use the latest from Oracle instead.
2. QZ Tray 2.0 will compile but just fail on HTML printing
3. QZ Tray 2.1 will not compile

So the question is really a matter of how to get the JavaFX libs installed and detected or disable them if they're not available.  I was able to reproduce the original bug report with the following steps:(for reference only, these don't work)
https://gist.github.com/tresf/a67fc559ef60cf077644175dc081397e

**Comment by tresf** [comment]
Ok... it turns out that `SwingFXUtils.java` is not part of the the Oracle JDK or the Gluon JDK, which is why compilation isn't possible on Raspberry Pi.

My best guess is this was done because it uses `IntegerComponentRaster` which is a proprietary API.  See the following compilation warning:

```
    [javac] /home/pi/tray/src/javafx/embed/swing/SwingFXUtils.java:38: warning:
       IntegerComponentRaster is internal proprietary API and may be removed in a future release
```

@Fermec28 can you try the following:

```bash
cd ~/tray
mkdir -p src/javafx/embed/swing
wget https://gist.githubusercontent.com/tresf/c50e3f7740b731a92519d8a98437c749/raw/SwingFXUtils.java -O src/javafx/embed/swing/SwingFXUtils.java
ant #... etc
```

From my env:

```ini
[...]
get-version:
     [echo] Version 2.1.0-RC4
     [echo] Size: 15418 KB
[...]
```

Unfortunately, this leads to even more runtime errors such as `java.lang.NoClassDefFoundError: javafx/print/PrintColor`.  This may be due to the class path of Gluon not being setup properly, so if there's any more insight as to get this future, please provide the instructions here. :)

```
java.lang.RuntimeException: Cannot call method public void qz.ws.PrintSocketClient#onMessage(org.eclipse.jetty.websocket.api.Session, java.io.Reader) with args: [org.eclipse.jetty.websocket.common.WebSocketSession, org.eclipse.jetty.websocket.common.message.MessageReader]
	at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.unwrapRuntimeException(CallableMethod.java:93)
	at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:75)
	at org.eclipse.jetty.websocket.common.events.annotated.OptionalSessionCallableMethod.call(OptionalSessionCallableMethod.java:68)
	at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver$2.run(JettyAnnotatedEventDriver.java:210)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NoClassDefFoundError: javafx/print/PrintColor
	at qz.printer.PrintOptions$ColorType.<clinit>(Unknown Source)
	at qz.printer.PrintOptions$Pixel.<init>(Unknown Source)
	at qz.printer.PrintOptions.<init>(Unknown Source)
	at qz.utils.PrintingUtilities.processPrintRequest(Unknown Source)
	at qz.ws.PrintSocketClient.processMessage(Unknown Source)
	at qz.ws.PrintSocketClient.onMessage(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:70)
	... 5 more
Caused by: java.lang.ClassNotFoundException: javafx.print.PrintColor
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 16 more

```

**Comment by tresf** [comment]
@Fermec28 / @blaggacao I'm closing this out as wontfix due to the very poor JavaFX support on Pi.  I think it's best to provide a way to disable compiling in JavaFX support, but I'm hesitant to add it since most environments (Mac, Windows, Linux) all have the JavaFX libraries present, so it's less of matter of being able to build it and more of a matter of WHERE to build it.

We'd welcome a PR with a way to detect or toggle this off for any interested parties.  Until then, it's `wontfix`.
